### PR TITLE
ENH: Outsource leave-one-out splitter so it can be used across data types

### PR DESCRIPTION
**Pull request by teresamg**
(Empty message edited by @oesteban)

This PR covers two targets:

- [x] Extract the "logo" splitter from the `DWI` object. The rationale behind this is allowing its use across different data types (e.g., with PET).
- [x] Incidentally, change the name from "logo" (for leave-one-gradient-out) to "lovo" (for leave-one-volume-out).

### CODE REVIEW

**Comment by oesteban** [comment]
After refactoring the package's structure, I have allowed myself to rebase into `main`.

I would suggest we pull the `logo_split` function out of the `dmri.py` submodule, and instead one of these two options:

- Under the `eddymotion.data` module, e.g. in `eddymotion.data.splitting`.
- Under the `eddymotion.model` module, which is closer to what scikit-learn has (I believe it is something along the lines of `sklearn.model.model_selection`). However, in our case this has nothing to do with model selection, but rather a data manipulation routine.

I would favor creating a module under `eddymotion.data` - whether it is `splitting.py`, I don't really care much.

WDYT?

**Comment by oesteban** [comment]
In e2b8f68 I went ahead and just started the `eddymotion.data.splitting` submodule. Please amend if you feel that's necessary.

@teresamg my rebasing will require you to update your working copy. If you don't have changes that you did not push, then I would just switch to main, `git branch -D teresamg` and then `git checkout teresamg`. If you did have changes, then let me know and I can help how to best transfer them.

File: src/eddymotion/data/dmri.py [hunk 1]
```diff
@@ -74,61 +74,16 @@ def __len__(self):
...
-                axis=-1,
-            )
-
-        return (
-            (train_data, train_gradients),
-            (dwframe, bframe),
-        )
-
-    def set_transform(self, index, affine, order=3):
+    def set_transform(self, dwframe, bvec, index, affine, order=3):
```
Comment (oesteban) [inline src/eddymotion/data/dmri.py:86]: This feels like trying to overcome the memory issues - let's address them separately.

File: src/eddymotion/estimator.py [hunk 2]
```diff
@@ -132,7 +136,12 @@ def fit(
                         pbar.set_description_str(
                             f"Pass {i_iter + 1}/{n_iter} | Fit and predict b-index <{i}>"
                         )
-                        data_train, data_test = dwdata.logo_split(i, with_b0=True)
+                        dwframe = np.asanyarray(dwdata.dataobj[..., i])
+                        bframe = np.asanyarray(dwdata.gradients[..., i])
+                        data_train, data_test = logo_split(
+                            dwdata, dwframe, bframe, i, with_b0=True
```
Comment (oesteban) [inline src/eddymotion/estimator.py:142]: Function signature should not change (except for making the data object the first argument and removing `with_b0` which someone (me) added on a gloomy day.

File: src/eddymotion/data/splitting.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+# We support and encourage derived works from this project, please read
+# about our expectations at
+#
+#     https://www.nipreps.org/community/licensing/
+#
+"""Data splitting helpers."""
+import numpy as np
+
+
+def lovo_split(dwdata, dwframe, bframe, index, with_b0=False):
```
Comment (oesteban) [inline src/eddymotion/data/splitting.py:27]:
```suggestion
def lovo_split(data, index):
```

File: src/eddymotion/data/splitting.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        Test 3D map (one DWI orientation) and corresponding b-vector/value.
+        The test data/gradient come **from the original dataset**.
+
+    """
+
+    # if the size of the mask does not match data, cache is stale
+    mask = np.zeros(len(dwdata), dtype=bool)
+    mask[index] = True
+
+    train_data = dwdata.dataobj[..., ~mask]
```
Comment (oesteban) [inline src/eddymotion/data/splitting.py:55]:
```suggestion
    train_data = data.dataobj[..., ~mask]
```

File: src/eddymotion/data/splitting.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    train_gradients = dwdata.gradients[..., ~mask]
```
Comment (oesteban) [inline src/eddymotion/data/splitting.py:56]:
```suggestion
    train_gradients = data.gradients[..., ~mask]
```

File: src/eddymotion/data/splitting.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            (np.asanyarray(dwdata.bzero)[..., np.newaxis], train_data),
+            axis=-1,
+        )
+        b0vec = np.zeros((4, 1))
+        b0vec[0, 0] = 1
+        train_gradients = np.concatenate(
+            (b0vec, train_gradients),
+            axis=-1,
+        )
+
```
Comment (oesteban) [inline src/eddymotion/data/splitting.py:69]: I will admit this is horrible. Can we get rid of this very bad idea (of mine)?
```suggestion
```

File: src/eddymotion/data/splitting.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    return (
+        (train_data, train_gradients),
+        (dwframe, bframe),
```
Comment (oesteban) [inline src/eddymotion/data/splitting.py:72]:
```suggestion
        (data.dataobj[..., mask], data.gradients[..., mask]),
```

File: src/eddymotion/data/splitting.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+import numpy as np
+
+
+def lovo_split(dwdata, dwframe, bframe, index, with_b0=False):
+    """
+    Produce one fold of LOVO (leave-one-volume-out).
+
+    Parameters
+    ----------
+    dwdata : :obj:`DWI`
```
Comment (oesteban) [inline src/eddymotion/data/splitting.py:33]:
```suggestion
    data : :obj:`eddymotion.data.dmri.DWI`
```

Comment (oesteban) [reply src/eddymotion/data/splitting.py:33]: this will need to be generalized afterward

File: src/eddymotion/data/splitting.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        DWI object
+    index : :obj:`int`
+        Index of the DWI orientation to be left out in this fold.
+    with_b0 : :obj:`bool`
+        Insert the *b=0* reference at the beginning of the training dataset.
```
Comment (oesteban) [inline src/eddymotion/data/splitting.py:38]:
```suggestion
```

**Comment by arokem** [comment]
@oesteban : This integrates all of your proposed changes, and updates things so that everything runs. It does also include the refactor of the `em_affines` list into a preallocated array, which helps with memory.

**Comment by oesteban** [comment]
I have resolved the conflicts and dropped some changes that did not pertain to this PR. These spurious changes addressed ancient memory issues which have been addressed elsewhere (namely, fixed within nitransforms, which was the original cause of the issues).

@esavary this is ready for your code review and perhaps some hands-on testing. Since I assume it is not immediate how to test eddymotion locally, I think we could add some documentation about that. Let us know how you want to proceed.

File: src/eddymotion/estimator.py [hunk 10]
```diff
@@ -150,9 +151,7 @@ def fit(
                         pbar.set_description_str(
                             f"Pass {i_iter + 1}/{n_iter} | Fit and predict b-index <{i}>"
                         )
-                        data_train, data_test = dwdata.logo_split(i, with_b0=True)
-                        grad_str = f"{i}, {data_test[1][:3]}, b={int(data_test[1][3])}"
-                        pbar.set_description_str(f"[{grad_str}], {n_jobs} jobs")
+                        data_train, data_test = logo_split(dwdata, i)
```
Comment (oesteban) [inline src/eddymotion/estimator.py:154]: Accidentaly removed these two lines, adding back:
```suggestion
                        data_train, data_test = logo_split(dwdata, i)
                        grad_str = f"{i}, {data_test[1][:3]}, b={int(data_test[1][3])}"
                        pbar.set_description_str(f"[{grad_str}], {n_jobs} jobs")
```

File: src/eddymotion/estimator.py [hunk 11]
```diff
@@ -33,6 +33,7 @@
 from pkg_resources import resource_filename as pkg_fn
 from tqdm import tqdm
 
+from eddymotion.data.splitting import lovo_split as logo_split
```
Comment (oesteban) [inline src/eddymotion/estimator.py:36]:
```suggestion
from eddymotion.data.splitting import lovo_split
```

File: src/eddymotion/estimator.py [hunk 12]
```diff
@@ -150,7 +151,7 @@ def fit(
                         pbar.set_description_str(
                             f"Pass {i_iter + 1}/{n_iter} | Fit and predict b-index <{i}>"
                         )
-                        data_train, data_test = dwdata.logo_split(i, with_b0=True)
+                        data_train, data_test = logo_split(dwdata, i)
```
Comment (oesteban) [inline src/eddymotion/estimator.py:154]:
```suggestion
                        data_train, data_test = lovo_split(dwdata, i)
```

File: src/eddymotion/data/dmri.py [hunk 13]
```diff
@@ -74,60 +74,6 @@ def __len__(self):
...
-
-        """
-        if not Path(self._filepath).exists():
-            self.to_filename(self._filepath)
-
-        # read original DWI data & b-vector
-        with h5py.File(self._filepath, "r") as in_file:
-            root = in_file["/0"]
-            dwframe = np.asanyarray(root["dataobj"][..., index])
-            bframe = np.asanyarray(root["gradients"][..., index])
```
Comment (oesteban) [inline src/eddymotion/data/dmri.py:105]: Access to the HDF5 file is missing in the outsourced version.

File: src/eddymotion/data/dmri.py [hunk 14]
```diff
@@ -74,60 +74,6 @@ def __len__(self):
...
-
-        # if the size of the mask does not match data, cache is stale
-        mask = np.zeros(len(self), dtype=bool)
-        mask[index] = True
-
-        train_data = self.dataobj[..., ~mask]
-        train_gradients = self.gradients[..., ~mask]
-
-        if with_b0:
```
Comment (oesteban) [inline src/eddymotion/data/dmri.py:114]: This argument has also disappeared in the outsourced version.

**Review by oesteban — COMMENTED** [review]
I've marked two important features that have disappeared in the outsourced function. @esavary - may you take care of those issues?

**Comment by esavary** [comment]
> I have resolved the conflicts and dropped some changes that did not pertain to this PR. These spurious changes addressed ancient memory issues which have been addressed elsewhere (namely, fixed within nitransforms, which was the original cause of the issues).
> 
> @esavary this is ready for your code review and perhaps some hands-on testing. Since I assume it is not immediate how to test eddymotion locally, I think we could add some documentation about that. Let us know how you want to proceed.

I was thinking of creating a 'test_splitting.py' module to test the splitting functionality. The idea is to generate a mock DWI (Diffusion-Weighted Imaging) object with 'dataobj' and 'gradients' attributes filled with zeros, except at one index where we'll set the values to 1. ( In the same direction as what you suggested yesterday) Then checking if the 'lovo_split' function, applied at this index, produces the expected results. Does this align with your thoughts?

Regarding documentation, could you provide more details? Do you expect a dedicated section focusing only on testing the splitting feature or more general about testing the entire module?

**Comment by esavary** [comment]
> After refactoring the package's structure, I have allowed myself to rebase into `main`.
> 
> I would suggest we pull the `logo_split` function out of the `dmri.py` submodule, and instead one of these two options:
> 
>     * Under the `eddymotion.data` module, e.g. in `eddymotion.data.splitting`.
> 
>     * Under the `eddymotion.model` module, which is closer to what scikit-learn has (I believe it is something along the lines of `sklearn.model.model_selection`). However, in our case this has nothing to do with model selection, but rather a data manipulation routine.
> 
> 
> I would favor creating a module under `eddymotion.data` - whether it is `splitting.py`, I don't really care much.
> 
> WDYT?

I've gone ahead and included the test I used to check the 'lovo_split' function in the test folder. However, feel free to remove it if you don't find it relevant or if you had something else in mind @oesteban .

File: src/eddymotion/data/splitting.py [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+        root = in_file["/0"]
+        dwframe = np.asanyarray(root["dataobj"][..., index])
+        bframe = np.asanyarray(root["gradients"][..., index])
+
+    # if the size of the mask does not match data, cache is stale
+    mask = np.zeros(len(data), dtype=bool)
+    mask[index] = True
+
+    train_data = data.dataobj[..., ~mask]
+    train_gradients = data.gradients[..., ~mask]
```
Comment (oesteban) [inline src/eddymotion/data/splitting.py:65]:
```suggestion
    train_data = data[..., ~mask]
    train_gradients = gradients[..., ~mask]
```

File: src/eddymotion/data/splitting.py [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+#
+#     https://www.nipreps.org/community/licensing/
+#
+"""Data splitting helpers."""
+from pathlib import Path
+import numpy as np
+import h5py
+
+
+def lovo_split(data, index, with_b0=False):
```
Comment (oesteban) [inline src/eddymotion/data/splitting.py:29]:
```suggestion
def lovo_split(dataset, index, with_b0=False):
```

File: src/eddymotion/data/splitting.py [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+        Training DWI and corresponding gradients.
+        Training data/gradients come **from the updated dataset**.
+    (test_data, test_gradients) :obj:`tuple`
+        Test 3D map (one DWI orientation) and corresponding b-vector/value.
+        The test data/gradient come **from the original dataset**.
+
+    """
+
+    if not Path(data.get_filename()).exists():
+        data.to_filename(data.get_filename())
```
Comment (oesteban) [inline src/eddymotion/data/splitting.py:52]:
```suggestion
    if not Path(dataset.get_filename()).exists():
        dataset.to_filename(data.get_filename())
```

File: src/eddymotion/data/splitting.py [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+
+    # read original DWI data & b-vector
+    with h5py.File(data.get_filename(), "r") as in_file:
```
Comment (oesteban) [inline src/eddymotion/data/splitting.py:55]:
```suggestion
    with h5py.File(dataset.get_filename(), "r") as in_file:
```

File: src/eddymotion/data/splitting.py [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+    # if the size of the mask does not match data, cache is stale
+    mask = np.zeros(len(data), dtype=bool)
+    mask[index] = True
+
+    train_data = data.dataobj[..., ~mask]
+    train_gradients = data.gradients[..., ~mask]
+
+    if with_b0:
+        train_data = np.concatenate(
+            (np.asanyarray(data.bzero)[..., np.newaxis], train_data),
```
Comment (oesteban) [inline src/eddymotion/data/splitting.py:69]:
```suggestion
            (np.asanyarray(dataset.bzero)[..., np.newaxis], train_data),
```

**Review by oesteban — COMMENTED** [review]
I'm doubtful about the right way to implement this. I'm not sure anymore if we want to access the filesystem here. In any case, I think the above variable changes will clarify the picture.

cc @effigies - what do you think? this is related to the new issue #145.

**Review by esavary — APPROVED** [review]

File: src/eddymotion/estimator.py [hunk 20]
```diff
@@ -150,7 +151,7 @@ def fit(
                         pbar.set_description_str(
                             f"Pass {i_iter + 1}/{n_iter} | Fit and predict b-index <{i}>"
                         )
-                        data_train, data_test = dwdata.logo_split(i, with_b0=True)
+                        data_train, data_test = lovo_split(dwdata, i)
```
Comment (oesteban) [inline src/eddymotion/estimator.py:154]:
```suggestion
                        data_train, data_test = lovo_split(dwdata, i, with_b0=True)
```

**Comment by arokem** [comment]
I think that @teresamg also had some test scripts that she used in order to run experiments with a sample dataset (I think that it was the traveling brain study from OpenNeuro). Maybe she can share those scripts (e.g., on a separate repo?) so that @esavary can run those for profiling/testing of this code?

**Comment by esavary** [comment]
> I think that @teresamg also had some test scripts that she used in order to run experiments with a sample dataset (I think that it was the traveling brain study from OpenNeuro). Maybe she can share those scripts (e.g., on a separate repo?) so that @esavary can run those for profiling/testing of this code?

Thanks, that sounds great! I'll run them as soon as I have them.
